src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_SRC_POS = 0;
+    private static final double GROW_SIZE = 1.5;
+    private T[] arrayData;
+    private int size;
+    private int fullSizeOfArray = 10;
+
+    @SuppressWarnings({"unchecked"})
+    public ArrayList() {
+        arrayData = (T[]) new Object[fullSizeOfArray];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        arrayData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        isIndexLegal(index);
+        ensureCapacity();
+        arrayCopyAdd(index);
+        arrayData[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexGetLegal(index);
+        return arrayData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexGetLegal(index);
+        arrayData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexGetLegal(index);
+        T value = arrayData[index];
+        arrayCopyRemove(index);
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((element == null
+                    && arrayData[i] == null)
+                    || element != null && element.equals(arrayData[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element are not exist in data");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size <= 0;
+    }
+
+    @SuppressWarnings({"unchecked"})
+    private void grow() {
+        T[] arrayDataTemp = (T[]) new Object[(int) (fullSizeOfArray * GROW_SIZE)];
+        System.arraycopy(arrayData, INITIAL_SRC_POS,
+                arrayDataTemp, INITIAL_SRC_POS, arrayData.length);
+        arrayData = arrayDataTemp;
+        fullSizeOfArray = (int) (fullSizeOfArray * GROW_SIZE);
```

Comment:
we don't need to store this data in a separate variable as we can get in from `arrayData.length`